### DISCUSSION TITLE
Trying to work with llama3 in Azure using AzureMLChatOnlineEndpoint but no cost showing up

### DISCUSSION THREAD

**Discussion by spacepirate0001** [langfuse/langfuse#2431] [Support]
working with Llama3 in Azure I am using AzureMLChatOnlineEndpoint to define llm for runnable_chain but this function does not support azure_deployment or api_version like AzureChatOpenAI. I have defined a model in LangFuse but cost is not being relfected.

**Comment by marcklingen** [comment]
do you see correct token counts in Langfuse?

**Comment by spacepirate0001** [comment]
I don’t!

**Reply by spacepirate0001** [reply]
Maybe an example of working with self hosted models using AzureMLChatOnlineEndpoint would help.
Here is how I added my model to the model list:
![image](https://github.com/langfuse/langfuse/assets/6020019/3d65958d-a667-4e99-a4f0-007d3f52f58b)

I also noticed that the tool rounds the input and output prices which might not be accurate. I entered $0.00000378 and $0.00001134 but it rounded them up to what is in the image. Finally. even End Time is not recorded either when using llama3 in Azure vs gpt4.
![image](https://github.com/langfuse/langfuse/assets/6020019/f06d4fe6-4468-495e-899a-0aa297d90790)

**Reply by spacepirate0001** [reply]
I got some error in log! Seems like LangFuse is not handling the llm response correctly:

`WARNING:langfuse:Langfuse was not able to parse the LLM model. The LLM call will be recorded without model name. Please create an issue so we can fix your integration: https://github.com/langfuse/langfuse/issues/new/choose
ERROR:langfuse:Got unknown type content='{<MY LLM DATA RESPONSE>}' response_metadata={'finish_reason': 'stop', 'logprobs': None} type='assistant' id='run-081ef404-821a-4779-bb6e-8b0407564bee-0'
Traceback (most recent call last):
  File ".venv\lib\site-packages\langfuse\callback\langchain.py", line 719, in on_llm_end
    self._convert_message_to_dict(generation.message)
  File ".venv\lib\site-packages\langfuse\callback\langchain.py", line 838, in _convert_message_to_dict
    raise ValueError(f"Got unknown type {message}")
ValueError: Got unknown type content='{<MY LLM DATA RESPONSE>}' response_metadata={'finish_reason': 'stop', 'logprobs': None} type='assistant' id='run-081ef404-821a-4779-bb6e-8b0407564bee-0'`

**Reply by spacepirate0001** [reply]
Ok I found a few issues:
1-  The message passed in `_convert_message_to_dict` from llama3 is type `assistant` where the function is expecting type `ai` thus the error `ValueError(f"Got unknown type {message}")`, these types are imported from `langchain_core` so not sure how to address this.
2- In `_parse_usage(response: LLMResult) if response.llm_output is not None:` is looking for `llm.output` which is coming back `none` so no usage data to parse
3- in the same function `if hasattr(response, "generations"):` evaluates to `false` because the response has `generation_info` instead.

example response:
`LLMResult(generations=[[ChatGeneration(text='{<MY_CONTENT>}', generation_info={'finish_reason': 'stop', 'logprobs': None}, message=BaseMessage(content='{<MY_CONTENT>}', response_metadata={'finish_reason': 'stop', 'logprobs': None}, type='assistant', id='run-88b5b79a-68d4-4b50-bb71-235f2e1e40e9-0'))]], llm_output=None, run=None)`
